diff --git a/core/includes/install.core.inc b/core/includes/install.core.inc
index b4ec568004..de0fd9c01a 100644
--- a/core/includes/install.core.inc
+++ b/core/includes/install.core.inc
@@ -1169,23 +1169,19 @@ function install_verify_database_settings($site_path) {
  *   Thrown when the database already has a table that would be created by
  *   installing the System module.
  */
-function install_verify_database_ready() {
-  $system_schema = system_schema();
-  end($system_schema);
-  $table = key($system_schema);
-
-  $existing_install = FALSE;
-  if ($database = Database::getConnectionInfo()) {
-    try {
-      $existing_install = Database::getConnection()->schema()->tableExists($table);
-    }
-    // Do not trigger an error if the database query fails, since the database
-    // might not be set up yet.
-    catch (\Exception $e) {
+function install_verify_database_ready(&$install_state) {
+  // If the installation fails mid-install, we want the database to be empty on
+  // the next run.
+  if (empty($install_state['database_ready']) && empty($install_state['parameters']['db_cleared'])) {
+    $tables = Database::getConnection()->schema()->findTables('%');
+    if ($tables) {
+      foreach ($tables as $table) {
+        drupal_set_message($table);
+        Database::getConnection()->schema()->dropTable($table);
+      }
     }
-  }
-  if ($existing_install) {
-    throw new AlreadyInstalledException(\Drupal::service('string_translation'));
+    $install_state['database_ready'] = TRUE;
+    $install_state['parameters']['db_cleared'] = TRUE;
   }
 }
 
